File: StriverA2Z/Java/LinkedLists/Questions/SinglyLinkedList/ReverseLinkedList.java
public class ReverseLinkedList {
    public static void main(String[] args) {
        Node head = LinkedListOperations.arrayToLinkedList(new int[]{1, 3, 2, 8});
        Node reverseList = reverseNormal(head);
        LinkedListOperations.traverseLinkedList(reverseList);
        Node recursionReverse = reverseRecursion(head);
        LinkedListOperations.traverseLinkedList(recursionReverse);
    }

    static Node reverseNormal(Node head) {
        Node prev = null;
        Node temp = head;
        while (temp != null) {
            Node front = temp.next;
            temp.next = prev;
            prev = temp;
            temp = front;
        }
        return prev;
    }

    static Node reverseRecursion(Node head) {
        if (head == null || head.next == null) return head;
        Node newHead = reverseRecursion(head.next);
        Node front = head.next;
        front.next = head;
        head.next = null;
        return newHead;
    }
}

File: StriverA2Z/C++/6_LinkedList/Questions/SinglyLinkedList/reverse_linked_list.cpp
#include<bits/stdc++.h>
#include "Node.h"

using namespace std;

Node *reverseNormal(Node *head) {
    Node *temp = head;
    Node *prev = nullptr;
    while (temp) {
        Node *front = temp->next;
        temp->next = prev;
        prev = temp;
        temp = front;
    }
    return prev;
}

Node *reverseRecursion(Node *head) {
    if (head == nullptr || head->next == nullptr) {
        return head;
    }
    Node *newHead = reverseRecursion(head->next);
    Node *front = head->next;
    front->next = head;
    head->next = nullptr;
    return newHead;
}

int main() {
    vector<int> arr = {1, 3, 2, 8};
    Node *head = array_to_linkedlist(arr);
    Node *reverseList = reverseNormal(head);
    Node *recursiveReverse = reverseRecursion(head);
    traverseLinkedList(recursiveReverse);
    traverseLinkedList(reverseList);
    return 0;
}